sdk/src/main/java/io/opentdf/platform/sdk/nanotdf/ResourceLocator.java

```diff
@@ -1,37 +1,140 @@
+            } else if (identifierLen <= 2) {
+                this.identifierType = NanoTDFType.IdentifierType.TWO_BYTES;
+                this.identifier = new byte[NanoTDFType.IdentifierType.TWO_BYTES.getLength()];
+                System.arraycopy(identifier.getBytes(), 0, this.identifier, 0, identifierLen);
+            } else if (identifierLen <= 8) {
+                this.identifierType = NanoTDFType.IdentifierType.EIGHT_BYTES;
+                this.identifier = new byte[NanoTDFType.IdentifierType.EIGHT_BYTES.getLength()];
+                System.arraycopy(identifier.getBytes(), 0, this.identifier, 0, identifierLen);
+            } else if (identifierLen <= 32) {
+                this.identifierType = NanoTDFType.IdentifierType.THIRTY_TWO_BYTES;
+                this.identifier = new byte[NanoTDFType.IdentifierType.THIRTY_TWO_BYTES.getLength()];
+                System.arraycopy(identifier.getBytes(), 0, this.identifier, 0, identifierLen);
+            } else {
+                throw new IllegalArgumentException("Unsupported identifier length: " + identifierLen);
+            }
+        }
     }
 
     public ResourceLocator(ByteBuffer buffer) {
         // Get the first byte and mask it with 0xF to keep only the first four bits
-        byte protocolByte = buffer.get();
-        int protocolIndex = protocolByte & 0xF;
-        this.protocol = NanoTDFType.Protocol.values()[protocolIndex];
+        final byte protocolWithIdentifier = buffer.get();
+        int protocolNibble = protocolWithIdentifier & 0x0F;
+        int identifierNibble = (protocolWithIdentifier & 0xF0) >> 4;
+        this.protocol = NanoTDFType.Protocol.values()[protocolNibble];
+        // body
         this.bodyLength = buffer.get();
         this.body = new byte[this.bodyLength];
         buffer.get(this.body);
+        // identifier
+        this.identifierType = NanoTDFType.IdentifierType.values()[identifierNibble];
+        switch (this.identifierType) {
+            case NONE:
+                this.identifier = new byte[0];
+                break;
+            case TWO_BYTES:
+                this.identifier = new byte[2];
+                buffer.get(this.identifier);
+                break;
+            case EIGHT_BYTES:
+                this.identifier = new byte[8];
+                buffer.get(this.identifier);
+                break;
+            case THIRTY_TWO_BYTES:
+                this.identifier = new byte[32];
+                buffer.get(this.identifier);
+                break;
+            default:
+                throw new IllegalArgumentException("Unexpected identifier type: " + identifierType);
+        }
+    }
+
+    public void setIdentifier(String identifier) {
+        if (identifier == null) {
+            this.identifierType = NanoTDFType.IdentifierType.NONE;
+            this.identifier = new byte[0];
+        } else {
+            byte[] identifierBytes = identifier.getBytes();
```

Comment:
I think we should set the charset here. If we [don't it uses the platform default](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#getBytes--)